osh/string_ops.py

```diff
@@ -53,7 +55,95 @@ def _Utf8CharLen(starting_byte):
         e_strict(INVALID_START, loc.Missing)
 
 
-def _NextUtf8Char(s, i):
+def _ReadOneUnit(s, cursor):
+    # type: (str, int) -> Tuple[int, int]
+    """Helper for DecodeUtf8Char"""
+    if cursor >= len(s):
+        raise error.Expr(INCOMPLETE_CHAR, loc.Missing)
```

Comment:
Hm yeah I remember this location issue, the problem is basically parsing strings at runtime, and that has to be connected back to the source location somehow

I imagine the error location is not good now

Or I guess it would just be ideal if it points to the `(` in  `trim()`?

(I just did a huge refactoring of `test/ysh-runtime-errors.sh`, the file we use to test and eyeball the errors)

Reply:
It may also be nice to emit the offending byte offset. For example, in python 3.10:

```python
>>> bytes([0x20, 0xF9]).decode('utf8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf9 in position 1: invalid start byte
```

But, for now, I've decided to keep the error messages consistent with the other UTF-8 string operations.